### PR TITLE
add deviceInitializer and api request for it

### PR DESCRIPTION
**Pull request by MaxMagazin**
- add abstract DeviceInitializer class
- add api request for it (/api/v1/device/initialize)

### CODE REVIEW

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.java [hunk 1]
```diff
@@ -208,6 +244,9 @@ private ApiResponse apiCall(Request request, boolean json) throws ApiException,
             throw new ApiException("Server error: Resource not found.");
         } else if (response.code() == 304) {
             throw new ResourceNotModified();
+        } else if (response.code() == 400) {
+            response.close();
+            throw new ApiException("This initialization token has already been used.");
```
Comment (raphaelm) [inline libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.java]: I'm not sure if it makes sense to use the same postResource() here, as this message only makes sense for the initialization request, while an error 400 can be returned for a multitude of ohter reasons.

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/api/DeviceInitializer.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            String api_token = responseJson.getString("api_token");
+            String name = responseJson.getString("name");
+
+            InitializationResult result = new InitializationResult(organizer, device_id, unique_serial, api_token, name);
+            return result;
+
+        } catch (JSONException e) {
+            e.printStackTrace();
+        } catch (ApiException e) {
+            e.printStackTrace();
```
Comment (raphaelm) [inline libpretixsync/src/main/java/eu/pretix/libpretixsync/api/DeviceInitializer.java]: We should try to be able to pass a proper error message here.

Comment (MaxMagazin) [reply libpretixsync/src/main/java/eu/pretix/libpretixsync/api/DeviceInitializer.java]: String errror message will be enough? i.e. we could encapsulate it as String in InitializationResult object

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.java (hunk 1 above)
Comment (MaxMagazin) [reply libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.java]: you are right, will introduce separate method for that

**Comment by MaxMagazin** [comment]
have added a few changes, let me know what you think when you are back

**Comment by raphaelm** [comment]
Hi,

this looks good, I hope you still forgive me for not merging this, since I just pushed something very similar to the ``v2`` branch. The code I pushed has already been in use in our closed-source pretixPOS application, so I could just copy it and be relatively sure that it's well-tested and works.

FYI: We're currently preparing a successor to pretixdroid which will use the new API. We'll release it as open source as soon as it's working.

**Pull request closed without merging**